File: Part 3 - Classification/Section 14 - Logistic Regression/logistic_regression_me.R
dataset=read.csv('Social_Network_Ads.csv')
dataset=dataset[,3:5]

#Splitting the dataset into training and test set
#install.packages('caTools')

library(caTools)
set.seed(123)
split=sample.split(dataset$Purchased,SplitRatio = 0.75)
training_set=subset(dataset,split==TRUE)
test_set=subset(dataset,split==FALSE)

#Feature SCaling Excluding categorical values
training_set[,1:2]=scale(training_set[,1:2])
test_set[,1:2]=scale(test_set[,1:2])

#Fitting Logistic Regression to the Training Set
classifier = glm(formula = Purchased ~ .,
                 family = binomial,
                 data = training_set) # Doubt binomial family

#Predicting the Test set results
prob_pred=predict(classifier, type = 'response', newdata = test_set[-3])
y_pred=ifelse(prob_pred > 0.5, 1, 0)

# Making the confusion metrics (Number correct and incorrect predictions)
cm = table(test_set[,3],y_pred)

#Visualising the Training set results
#install.packages('ElemStatLearn')
#Point is the  truth and region is the prediction
#Straight line is called the prediction boundary and it is a staright line because
#Logistic Regression is Linear Classifier (here 2D).
#(in a 3D it will be a straight plane).
library(ElemStatLearn)
set=training_set
X1=seq(min(set[,1])-1,max(set[,1])+1, by = 0.01)
X2=seq(min(set[,2])-1,max(set[,2])+1, by = 0.01)
grid_set=expand.grid(X1,X2)
colnames(grid_set)=c('Age','EstimatedSalary')
prob_set = predict(classifier, type = 'response', newdata = grid_set)
y_grid = ifelse(prob_set > 0.5, 1, 0)
plot(set[,-3],
     main = 'Logistic Regression (Training Set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(X1), ylim = range(X2))
contour(X1,X2,matrix(as.numeric(y_grid),length(X1),length(X2)),add = TRUE)
points(grid_set,pch = '.',col = ifelse(y_grid==1,'springgreen3','tomato'))
points(set,pch = 21,bg=ifelse(set[,3]==1,'green4','red3'))


#Visualising the Test set results
set=test_set
X1=seq(min(set[,1])-1,max(set[,1])+1, by = 0.01) #Making the pixel values
X2=seq(min(set[,2])-1,max(set[,2])+1, by = 0.01) #Same as Python
grid_set=expand.grid(X1,X2) #Now making the pixel grid for real as well as imaginary users.
colnames(grid_set)=c('Age','EstimatedSalary') #Giving name to the columns of matrix 
prob_set = predict(classifier, type = 'response', newdata = grid_set) #Predicting for grid_set matrix for each skills
y_grid = ifelse(prob_set > 0.5, 1, 0)
plot(set[,-3],
     main = 'Logistic Regression (Test Set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(X1), ylim = range(X2))
contour(X1,X2,matrix(as.numeric(y_grid),length(X1),length(X2)),add = TRUE) 
points(grid_set,pch = '.',col = ifelse(y_grid==1,'springgreen3','tomato')) #For Backgroound Predictions
points(set,pch = 21,bg=ifelse(set[,3]==1,'green4','red3')) #For true points

File: Part 3 - Classification/Section 14 - Logistic Regression/logistic_regression_me.py
# Data Preprocessing Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Social_Network_Ads.csv')
X = dataset.iloc[:, [2,3]].values
y = dataset.iloc[:, 4].values

# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

#Fitting Logistic Regression tothe Traning set
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state=0)
classifier.fit(X_train,y_train)

# Predicting The test set results
y_pred=classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred) #Upper Left Down Right shows correct
# Upper Right Down Left shows the wrong.

#Visualising the Training set results 
from matplotlib.colors import ListedColormap
X_set, y_set = X_train, y_train
X1, X2 = np.meshgrid(np.arange(start=X_set[:,0].min()-1,stop=X_set[:,0].max()+1,step=0.01),
                     np.arange(start=X_set[:,1].min()-1,stop=X_set[:,1].max()+1,step=0.01)) 
plt.contourf(X1,X2,classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
            alpha=0.75, cmap=ListedColormap(('red','green')))
plt.xlim(X1.min(),X1.max())
plt.ylim(X2.min(),X2.max())
for i,j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j,0], X_set[y_set == j,1],
                c = ListedColormap(('red','green'))(i), label=j)
plt.title('Logistic Regression (Traning Set)')
plt.xlabel('Age')
plt.ylabel('Estimatted Salary')
plt.legend()
plt.show()

#Visualising the Test set results 
from matplotlib.colors import ListedColormap
X_set, y_set = X_test, y_test #Just creating local variables.
X1, X2 = np.meshgrid(np.arange(start=X_set[:,0].min()-1,stop=X_set[:,0].max()+1,step=0.01),
                     np.arange(start=X_set[:,1].min()-1,stop=X_set[:,1].max()+1,step=0.01)) 
# It is actually specifying the pixel values with 0.01 gap.Now we dont want to squeeze the
#points to the corners so we take min-1 and max+1
plt.contourf(X1,X2,classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
            alpha=0.75, cmap=ListedColormap(('red','green')))
# Applying the classifiers on all the pixel classifiers observation points colurs 
# all the red and green pixel points. We use contour function to make the contour line.
# then we use predict function to predict each of the pixel point is in class0 or class 1
# and colouring them accordingly.
plt.xlim(X1.min(),X1.max()) # putting x limits
plt.ylim(X2.min(),X2.max()) # putting y limits
for i,j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j,0], X_set[y_set == j,1],
                c = ListedColormap(('red','green'))(i), label=j)
# plotting all the points in scatter plot.    
plt.title('Logistic Regression (Test Set)')
plt.xlabel('Age')
plt.ylabel('Estimatted Salary')
plt.legend()
plt.show()




















